Add automatic stylesheet enqueuing

lukecarbis opened the pull request:
This is a _much_ neater way to do block styles.

On the frontend, Block Lab will enqueue the first of these stylesheets, for any blocks used on the current page:
- `{child theme directory}/blocks/css/block-{block name}.css`
- `{child theme directory}/blocks/block-{block name}.css`
- `{parent theme directory}/blocks/css/block-{block name}.css`
- `{parent theme directory}/blocks/block-{block name}.css`

In the editor, Block Lab will enqueue the first of these stylesheets, for _all blocks_ (since any one of them could be inserted at any point):
- `{child theme directory}/blocks/css/preview-{block name}.css`
- `{child theme directory}/blocks/preview-{block name}.css`
- `{child theme directory}/blocks/css/block-{block name}.css`
- `{child theme directory}/blocks/block-{block name}.css`
- `{parent theme directory}/blocks/css/preview-{block name}.css`
- `{parent theme directory}/blocks/preview-{block name}.css`
- `{parent theme directory}/blocks/css/block-{block name}.css`
- `{parent theme directory}/blocks/block-{block name}.css`

Please test each possibility!

Commit:
Add automatic stylesheet enqueuing

RobStino commented:
:slow_clap:
This is so cool man! Such an elegant solution (to be tested) to styling for blocks!
Bonus points for the conditional enqueueing on the frontend!

Commit:
More gracefully handle prepending a slash to the stylesheet url

lukecarbis commented:
Do not merge until I've written unit tests, which I'll do once #233 is merged.

Commit:
Simplify function call to block_lab_locate_template

Commit:
Merge branch 'develop' into feature/auto-enqueue-css

Commit:
Add unit test for block_lab_enqueue_styles

lukecarbis commented:
Okay, @kienstra ready for you to review and merge.

RobStino commented:
Tested all variations of file locations and everything works great!

kienstra commented on php/blocks/class-loader.php:
```diff
@@ -207,17 +215,19 @@ public function render_block_template( $block, $attributes ) {
 				}
 			}
+
+			block_lab_enqueue_styles( $block['name'], 'block' );
```
Would this method fit well in the `Loader` class, instead of `php/helpers.php`? Though maybe it'll later be used in other classes, I'm not sure.

lukecarbis replied:
Great idea – done in 6148de0.

kienstra commented on tests/php/test-helpers.php:
```diff
@@ -38,4 +38,57 @@ public function test_block_field() {
+	 * @covers ::block_lab_enqueue_styles()
+	 */
+	public function test_block_lab_enqueue_styles() {
+		global $wp_styles;
```
Sometimes I'll use `$wp_styles = wp_styles();`, but the effect is the same.

lukecarbis replied:
Updated.

kienstra reviewed (dismissed):
**Approved**

Hi @lukecarbis,
Great idea to enqueue styling for the blocks. The styling appeared as expected:

<img width="745" alt="styling-appears" src="https://user-images.githubusercontent.com/4063887/54570706-1eff7380-49a5-11e9-9b39-e59444e84ff1.png">

There's a question below, but not a blocker to merging if you're satisfied.

RobStino reviewed (dismissed):
Tested with all file variations and all worked as expected. Great work!

Commit:
Move block_lab_template_part and block_lab_enqueue_styles into the Loader class

lukecarbis commented:
@kienstra Could you please review 6148de0?

kienstra committed:
Change name of function to the method

block_lab_enqueue_styles() was moved inside
class Loader.
So change the call to $this->enqueue_block_styles().

kienstra reviewed (approved):
**Approved**

Hi @lukecarbis,
Sorry for the delay. This looks good, and the block stylesheet still appears as expected:

![stylesheet-enqueued](https://user-images.githubusercontent.com/4063887/54729582-04b2ca80-4b4a-11e9-855d-226599ca746b.jpg)
